File: learn/algo/Algo.py
#Created by Trotiflex https://github.com/Trotiflex

import math

L1 = 5 #coxa
L2 = 6 #femur
L3 = 13.5 #tibia

x = 13 #Extension de la patte
z = 8 #Hauteur

def Racine(a, b):
    return math.sqrt((a*a) + (b*b))

def AlKashi(a, b, c):
    return math.acos((a*a +b*b - c*c) / (2*a*b))

def RadToDegrees(a):
    return a*(180/math.pi)

Hypotenuse = Racine(x, z)
print("Hypotenuse:",Hypotenuse)

A1 = AlKashi(z, Hypotenuse, x)
A2 = AlKashi(Hypotenuse, L2, L3)
A3 = AlKashi(L2, L3, Hypotenuse)

AngleFemur = 180 - (RadToDegrees(A1) + RadToDegrees(A2))

print("L'angle du femur est:", AngleFemur)

File: software/Arduino/src/functions/Algo.cpp
#include "Algo.h"
#include "HexaMath.h"
#include "../config/config.h"
#include "math.h"
#include "Arduino.h"

LegAngles Algo(float x, float z, float y, int duree) {
    LegAngles angles;
    /*
    x = taille de la patte; z = hauteur; y = rotation (cm);
    duree = durée du mouvement en millisecondes;
    */
   //Calcul des valeurs manquantes ( Taille de la patte vu du ciel, et la taille de l'hypotenuse formé par le bout du coxa et çje bout de la patte )
    float TPatte = Racine(y, x);
    float hyp = Racine(z, TPatte-coxa);

    float A1 = radiansToDegrees(atan((TPatte-coxa)/z));
    float A2 = AlKashi(femur, hyp, tibia);
    angles.AngleFemur = A1 + A2;

    float A3 = AlKashi(tibia, femur, hyp);
   
    // Serial.println("\nTibia: ");
    // Serial.println(A3);

    angles.AngleCoxa = radiansToDegrees(atan(y / x)) + 90;
    angles.AngleTibia = A3;

    return angles;
}
